feat(FI-961) add ICRC-1 index tutorial (#1937)
dfinity/portal: docs/developer-docs/integrations/icrc-1/icrc1-index-setup.md

      "wasm": "https://download.dfinity.systems/ic/d87954601e4b22972899e9957e800406a0a6b929/canisters/ic-icrc1-ledger.wasm.gz",
    }
  },
  "output_env_file": ".env",
  "version": 1
}
```
The data for the ICRC-1 ledger were also added. You can look at the [ICRC-1 ledger deployment guide](./icrc1-ledger-setup.md) to see why.
In an existing project you would only need to add the `icrc1_index_canister` and `icrc1_ledger_canister` canisters to the `canisters` section.

### Step 5 [Optional]:  Start a local replica.
This step can be skipped if you already have a local replica up and running. 
``` sh
dfx start --background --clean
```

### Step 6: Deploy the ICRC-1 index canister:
Here it is assumed that the canister id of your local ICRC-1 ledger is `mxzaz-hqaaa-aaaar-qaada-cai`, otherwise replace it with your ICRC-1 ledger canister id. 

```
dfx deploy icrc1_index_canister --argument '(opt variant{Init = record {ledger_id = principal "mxzaz-hqaaa-aaaar-qaada-cai"}})'
```

The ICRC-1 index canister will start synching right away and will automatically try to fetch new blocks from the ICRC-1 ledger every few seconds. 

> ### Step 7: Check the status and ICRC-1 ledger id on the ICRC-1 index canister.

Reviewer: ```suggestion
### Step 7: Check the status and ICRC-1 ledger ID on the ICRC-1 index canister.
```

> You can check that the correct ledger id was set but runnig the following command.

Reviewer: ```suggestion
You can check that the correct ledger ID was set by running the following command.
```

```
dfx canister call icrc1_index_canister ledger_id '()'
```

> The result is the ledger canister id that the index canister is using to sync.

Reviewer: ```suggestion
The result is the ledger canister ID that the index canister is using to sync.
```

```
(principal "mxzaz-hqaaa-aaaar-qaada-cai")
```

```

> dfx canister call icrc1_index_canister status '()'

Reviewer: Should there be some description of what this command is doing?

```
It should return something like this:
```
(record { num_blocks_synced = 1 : nat64 })
```

> Depending on how many mint operations you created while setting up your ICRC-1 ledger the number of synched blocks here will be 0 if no initial balances were parsed or `X` if `X` initial balances were parsed. In the case of this tutorial the guide on setting up a local ledger was used and there only one initial balance was parsed as an initialization argument. Hence, the number of blocks synched at this stage is 1. 

Reviewer: ```suggestion
Depending on how many mint operations you created while setting up your ICRC-1 ledger, the number of synched blocks here will be 0 if no initial balances were parsed, or `X` if `X` initial balances were parsed. In the case of this tutorial, the guide on setting up a local ledger was used and there only one initial balance was parsed as an initialization argument. Hence, the number of blocks synched at this stage is 1. 
```

### Step 8: Create some new blocks to sync.

> You can check that the synchronization of the index is working by creating some transaction on the ICRC-1 ledger and then checking the status.

Reviewer: ```suggestion
You can check that the synchronization of the index is working by creating a transaction on the ICRC-1 ledger and then checking the status of that transaction.
```

> If you followed the guide on setting up an ICRC-1 ledger locally your default identity should have some ICRC-1 to be send. 

Reviewer: ```suggestion
If you followed the guide on setting up an ICRC-1 ledger locally, your default identity should have some ICRC-1 to be send. 
```

> Send some ICRC-1 to any prinICRC-1al.

Reviewer: ```suggestion
Send some ICRC-1 to any principal.
```

```
dfx canister call icrc1_ledger_canister icrc1_transfer '(record { to = record { owner = principal "npki3-wdfh4-siaeq-orwh4-bh5of-r7mxr-i35lm-6f2eh-rtmwp-dmzmn-tae";};  amount = 100000:nat;})'
```

```
dfx canister call icrc1_index_canister status '()'
```
It should now indicate that an additional block was synced compared to the last time we called the status endpoint.
```
(record { num_blocks_synced = 2 : nat64 })
```

> ### Step 9: Fetch some blocks

Reviewer: ```suggestion
### Step 9: Fetch some blocks.
```

You can use the ICRC-1 index canister to fetch blocks like so.
You have to specify the block at which you want to start fetching from (i.e. the lowest index you want to fetch). If you want to start from the beginning you have to set start to 0. Similarly, the length parameter indicates the number of blocks you would like to fetch. Since the last status call indicated that there are two blocks that were synced you can set this to 2. Note that if you specify more than 2 blocks it will simply return the maximum number of blocks the index contains (The limit of blocks per call is usually set to 2000 blocks).
```
dfx canister call icrc1_index_canister get_blocks '(record{start=0:nat;length=2:nat})'
```

> Which will return a vector the encoded blocks.

Reviewer: ```suggestion
Which will return a vector the encoded blocks:
```

                  variant { Int = 1_695_375_155_855_058_000 : int };
                };
              }
            };
          };
        }
      };
      variant {
        Map = vec {
          record {
            "phash";
            variant {
              Blob = blob "6\f0\02\d0\b0?3x\0b!\18e2\a6)h\b2z\82\cfz\0f\d5\ec>C\e4\05\a3\e0{\e6"
            };
          };
          record { "ts"; variant { Int = 1_695_375_453_656_274_000 : int } };
          record {
            "tx";
            variant {
              Map = vec {
                record { "amt"; variant { Int = 100_000 : int } };
                record { "op"; variant { Text = "mint" } };
                record {
                  "to";
                  variant {
                    Array = vec {
                      variant { Blob = blob "e?$\80\12\0e\8d\8f\c0\9f\ae,~\cb\c5\1b\ea\d9\e2\e8\87\8c\d9g\8d\99cf\02" };
                    }
                  };
                };
              }
            };
          };
        }
      };
    };
    chain_length = 2 : nat64;
  },
)
```

To fetch a vector of all transactions your default account was involved in you can use the following commands.

> Find out the principal of your default account.

Reviewer: ```suggestion
To find out the principal of your default account:
```

```
dfx identity get-principal
```
In the case of this tutorial this returns

```
hdq6b-ncywm-yajd5-4inc6-hgpzp-55xnp-py7d5-uqt6o-cv5c6-rrhwa-zqe
```

> Then you can query the transactions for this principal with the default subaccount set by calling. 

Reviewer: ```suggestion
Then you can query the transactions for this principal with the default subaccount set by calling:
```

```
dfx canister call icrc1_index_canister get_account_transactions '(record{account=record {owner = principal "hdq6b-ncywm-yajd5-4inc6-hgpzp-55xnp-py7d5-uqt6o-cv5c6-rrhwa-zqe"}; max_results=2:nat})'
```

The result will include the intial mint operation as well as the transfer that we made.

```
(
  variant {
    Ok = record {
      balance = 10_000_000_000 : nat;
      transactions = vec {
        record {
          id = 0 : nat;
          transaction = record {
            burn = null;
            kind = "mint";
            mint = opt record {
              to = record {
                owner = principal "hdq6b-ncywm-yajd5-4inc6-hgpzp-55xnp-py7d5-uqt6o-cv5c6-rrhwa-zqe";
                subaccount = null;
              };
              memo = null;
              created_at_time = opt (1_695_375_155_855_058_000 : nat64);
              amount = 10_000_000_000 : nat;
            };
            approve = null;
            timestamp = 1_695_375_155_855_058_000 : nat64;
            transfer = null;
          };
        };
      };
      oldest_tx_id = opt (0 : nat);
    }
  },
)
```

> The ICRC-1 ledger uses AccountIdentifier which are a hashed version of Account (Principal and Subaccount) for privacy. This also means that the returned transactions will show accounts as the hash in bytes rather than the actual Accounts.

Reviewer: ```suggestion
The ICRC-1 ledger uses `AccountIdentifier` which are a hashed version of `Account` (`Principal` and `Subaccount`) for privacy. This also means that the returned transactions will show accounts as the hash in bytes rather than the actual `Accounts`.
```

### Step 2 [Optional]: Create a new project folder.
It is advised you use the same project folder that you created during the ICRC-1 ledger setup. Alternatively, you can create a new one for the ICRC-1 index canister using the following command:

```
dfx new icrc1_index_canister
cd icrc1_index_canister
``` 

**OPTIONAL:**
If you created a new project folder, you will either have to make sure the `dfx.json` file contains the correct data on the ICRC-1 ledger (as described in the [ICRC-1 ledger setup guide](./icrc1-ledger-setup.md) or you communicate with the ICRC-1 ledger from its project folder that contains the correct `dfx.json file`. If you try to communicate with the ICRC-1 ledger from a new project folder where the `dfx.json` file does not contain information on the ICRC-1 ledger canister, you will not be able to create transactions. 

> ### Step 3:  Fetch the ICRC-1 index wasm and candid files

Reviewer: ```suggestion
### Step 3:  Fetch the ICRC-1 index Wasm and Candid files.
```

Go to the [releases overview](https://dashboard.internetcomputer.org/releases) and copy the latest replica binary revision (IC_VERSION). At the time of writing, this is `d87954601e4b22972899e9957e800406a0a6b929`.

The URL for the IPC index Wasm module is `curl -o index.wasm.gz "https://download.dfinity.systems/ic/$IC_VERSION/canisters/ic-icrc1-index-ng.wasm.gz"`, so with the above revision it would be `https://download.dfinity.systems/ic/d87954601e4b22972899e9957e800406a0a6b929/canisters/ic-icrc1-index-ng.wasm.gz`.

The URL for the ICRC-1 index .did file is `curl -o index.did "https://raw.githubusercontent.com/dfinity/ic/$IC_VERSION/rs/rosetta-api/icrc1/index-ng/index-ng.did"`, so with the above revision it would be `curl -o index.did "https://raw.githubusercontent.com/dfinity/ic/d87954601e4b22972899e9957e800406a0a6b929/rs/rosetta-api/icrc1/index-ng/index-ng.did"`.

> ### Step 4: Configure the `dfx.json` file

Reviewer: ```suggestion
### Step 4: Configure the `dfx.json` file.
```

>   Open the `dfx.json` file in your project's directory. Add the ICRC-1 index canister data and insert the canister data for your ICRC-1 ledger. If you followed the guide on local ICRC-1 ledger setup and you used the same project folder for both the ICRC-1 ledger and ICRC-1 index, your dfx.json file should look like this:

Reviewer: ```suggestion
Open the `dfx.json` file in your project's directory. Add the ICRC-1 index canister data and insert the canister data for your ICRC-1 ledger. If you followed the guide on local ICRC-1 ledger setup and you used the same project folder for both the ICRC-1 ledger and ICRC-1 index, your `dfx.json` file should look like this:
```

``` json
{
  "canisters": {
    "icrc1_index_canister": {
      "type": "custom",
      "candid": "https://raw.githubusercontent.com/dfinity/ic/d87954601e4b22972899e9957e800406a0a6b929/rs/rosetta-api/icrc1/index-ng/index-ng.did",
      "wasm": "https://download.dfinity.systems/ic/d87954601e4b22972899e9957e800406a0a6b929/canisters/ic-icrc1-index-ng.wasm.gz",
    },
    "icrc1_ledger_canister": {
      "type": "custom",
      "candid": "https://raw.githubusercontent.com/dfinity/ic/d87954601e4b22972899e9957e800406a0a6b929/rs/rosetta-api/icrc1/ledger/ledger.did",
      "wasm": "https://download.dfinity.systems/ic/d87954601e4b22972899e9957e800406a0a6b929/canisters/ic-icrc1-ledger.wasm.gz",
    }
  },
  "defaults": {
    "build": {
      "args": "",
      "packtool": ""

>     }

Author: I removed this for ICRC-1 and ICP index (ICP index tutorial had this line as well)

  },
  "output_env_file": ".env",
  "version": 1
}
```

> The data for the ICRC-1 ledger were also added. You can look at the [ICRC-1 ledger deployment guide](./icrc1-ledger-setup.md) to see why.

Reviewer: ```suggestion
The data for the ICRC-1 ledger was also added. You can look at the [ICRC-1 ledger deployment guide](./icrc1-ledger-setup.md) to see why.
```

In an existing project you would only need to add the `icrc1_index_canister` and `icrc1_ledger_canister` canisters to the `canisters` section.

### Step 5 [Optional]:  Start a local replica.
This step can be skipped if you already have a local replica up and running. 
``` sh
dfx start --background --clean
```

### Step 6:  Use your default identity to deploy the index canister.

``` sh
dfx identity use default
export DEFAULT=$(dfx identity get-principal)
```

> ### Step 7: Deploy the ICRC-1 index canister:

Reviewer: ```suggestion
### Step 7: Deploy the ICRC-1 index canister.
```

> Here it is assumed that the canister id of your local ICRC-1 ledger is `mxzaz-hqaaa-aaaar-qaada-cai`, otherwise replace it with your ICRC-1 ledger canister id. 

Reviewer: ```suggestion
Here it is assumed that the canister ID of your local ICRC-1 ledger is `mxzaz-hqaaa-aaaar-qaada-cai`, otherwise replace it with your ICRC-1 ledger canister ID. 
```
